tests/sharding/mainchain_handler/test_vmc_handler.py

```diff
@@ -0,0 +1,303 @@
+    return valcode_addr
+
+
+def deploy_initiating_contracts(vmc_handler, chain_handler, privkey):
+    if not is_vmc_deployed(vmc_handler, chain_handler):
+        address = privkey.public_key.to_canonical_address()
+        chain_handler.unlock_account(address, PASSPHRASE)
+        nonce = chain_handler.get_nonce(address)
+        txs = vmc_utils.mk_initiating_contracts(privkey, nonce, SpuriousDragonTransaction)
+        for tx in txs[:3]:
+            chain_handler.direct_tx(tx)
+        chain_handler.mine(1)
+        for tx in txs[3:]:
+            chain_handler.direct_tx(tx)
+            chain_handler.mine(1)
+        logger.debug(
+            'deploy_initiating_contracts: vmc_tx_hash=%s',
+            chain_handler.get_transaction_receipt(encode_hex(txs[-1].hash)),
+        )
+
+
+def import_key_to_mainchain_handler(chain_handler, key):
+    """
+    :param vmc_handler: VMCHandler
+    :param privkey: PrivateKey object from eth_keys
+    """
+    try:
+        chain_handler.w3.personal.importRawKey(key.to_hex(), PASSPHRASE)
+    # Exceptions happen when the key is already imported.
+    #   - ValueError: `web3.py`
+    #   - ValidationError: `eth_tester`
+    except (ValueError, ValidationError):
+        pass
+
+
+def get_testing_colhdr(vmc_handler,
+                       chain_handler,
+                       shard_id,
+                       parent_collation_hash,
+                       number,
+                       collation_coinbase=test_keys[0].public_key.to_canonical_address(),
+                       privkey=test_keys[0]):
+    period_length = PERIOD_LENGTH
+    expected_period_number = (chain_handler.get_block_number() + 1) // period_length
+    logger.debug("get_testing_colhdr: expected_period_number=%s", expected_period_number)
+    sender_addr = privkey.public_key.to_canonical_address()
+    period_start_prevhash = vmc_handler.call(
+        vmc_handler.mk_contract_tx_detail(sender_addr=sender_addr, gas=TX_GAS)
+    ).get_period_start_prevhash(expected_period_number)
+    logger.debug("get_testing_colhdr: period_start_prevhash=%s", period_start_prevhash)
+    tx_list_root = b"tx_list " * 4
+    post_state_root = b"post_sta" * 4
+    receipt_root = b"receipt " * 4
+    sighash = keccak(
+        rlp.encode([
+            shard_id,
+            expected_period_number,
+            period_start_prevhash,
+            parent_collation_hash,
+            tx_list_root,
+            collation_coinbase,
+            post_state_root,
+            receipt_root,
+            number,
+        ])
+    )
+    sig = vmc_utils.sign(sighash, privkey)
+    return rlp.encode([
+        shard_id,
+        expected_period_number,
+        period_start_prevhash,
+        parent_collation_hash,
+        tx_list_root,
+        collation_coinbase,
+        post_state_root,
+        receipt_root,
+        number,
+        sig,
+    ])
+
+
+def test_vmc_contract_calls(mainchain_handler):  # noqa: F811
```

Comment:
This is a very long test body.  At minimum I think it would benefit from adding as many comment blocks as possible that read as a narrative of what is happening and why we are doing it.

Reply:
Sure. Just want to confirm should I keep this long body, and just add enough comment blocks?
Because many of the steps depends on the previously made steps, like `deposit` + `mine(SHUFFLING_CYCLE_LENGTH)`. Originally because I test with two backends, `geth` and eth_tester, it takes too much time when testing with `geth`. So I compile the tests altogether. 
If we remove the logic of testing with `geth` from `test_vmc_handler.py`(thanks to your suggestion in the above), I can set the dependent steps as the fixture, and separate the long test body of `test_vmc_contract_calls` to multiple tests.

Reply:
Yes, for now we can keep the long body as long as it's well commented so that it's easier to follow.  I suspect we can trim this down into smaller pieces at some point but doesn't need to be now.